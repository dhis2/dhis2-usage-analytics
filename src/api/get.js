if (!process.env.REACT_APP_DHIS2_BASE_URL) {
  throw new Error('The environment variable REACT_APP_DHIS2_BASE_URL must be set');
}

const url = process.env.REACT_APP_DHIS2_BASE_URL;
const endpoint = `${url}/api`;
const defaultConfig = {
  method: 'GET',
  credentials: 'include',
}

export function get(path, config = defaultConfig) {
  if (!path) {
    return Promise.reject(new Error('A path must be passed'));
  }

  return fetch(`${endpoint}/${path}`, config)
    .then(response => {
      if (response.ok) {
        return response;
      }
    
      throw new Error(response.statusText)
    })
}

export function getJSON(path) {
  return get(path)
    .then(response => response.json())
    .then(data => {
      if (data.status === 'ERROR') {
        throw new Error(data.message)
      } else {
        return data
      }
    })
}
