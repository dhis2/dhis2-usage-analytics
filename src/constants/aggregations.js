import i18n from '@dhis2/d2-i18n'

export const SUM = 'SUM'
export const AVERAGE = 'AVERAGE'

const AGGREGATIONS = {
    [SUM]: { value: SUM, label: i18n.t('Sum') },
    [AVERAGE]: { value: AVERAGE, label: i18n.t('Average') },
}
export default AGGREGATIONS
