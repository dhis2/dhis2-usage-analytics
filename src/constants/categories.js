import i18n from '@dhis2/d2-i18n'

export const FAVORITE_VIEWS = 'FAVORITE_VIEWS'
export const FAVORITES_SAVED = 'FAVORITES_SAVED'
export const USERS = 'USERS'
export const TOP_FAVORITES = 'TOP_FAVORITES'
export const DATA_VALUES = 'DATA_VALUES'

const CATEGORIES = [
    { value: FAVORITE_VIEWS, label: i18n.t('Favorite views') },
    { value: FAVORITES_SAVED, label: i18n.t('Favorites saved') },
    { value: USERS, label: i18n.t('Users') },
    { value: TOP_FAVORITES, label: i18n.t('Top favorites') },
    { value: DATA_VALUES, label: i18n.t('Data values') },
]
export default CATEGORIES
