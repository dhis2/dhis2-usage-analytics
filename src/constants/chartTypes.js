import i18n from '@dhis2/d2-i18n'

export const ALL = 'ALL'
export const TOTAL = 'TOTAL'

const chartTypes = [
    { value: ALL, label: i18n.t('All') },
    { value: TOTAL, label: i18n.t('Total') },
]

export default chartTypes
