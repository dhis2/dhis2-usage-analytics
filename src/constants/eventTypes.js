import i18n from '@dhis2/d2-i18n'

export const CHART_VIEW = 'CHART_VIEW'
export const MAP_VIEW = 'MAP_VIEW'
export const REPORT_TABLE_VIEW = 'REPORT_TABLE_VIEW'
export const EVENT_REPORT_VIEW = 'EVENT_REPORT_VIEW'
export const EVENT_CHART_VIEW = 'EVENT_CHART_VIEW'
export const DASHBOARD_VIEW = 'DASHBOARD_VIEW'
export const DATA_SET_REPORT_VIEW = 'DATA_SET_REPORT_VIEW'

const EVENT_TYPES = {
    [CHART_VIEW]: { value: CHART_VIEW, label: i18n.t('Chart') },
    [MAP_VIEW]: { value: MAP_VIEW, label: i18n.t('Map') },
    [REPORT_TABLE_VIEW]: {
        value: REPORT_TABLE_VIEW,
        label: i18n.t('Report table'),
    },
    [EVENT_REPORT_VIEW]: {
        value: EVENT_REPORT_VIEW,
        label: i18n.t('Event report'),
    },
    [EVENT_CHART_VIEW]: {
        value: EVENT_CHART_VIEW,
        label: i18n.t('Event chart'),
    },
    [DASHBOARD_VIEW]: { value: DASHBOARD_VIEW, label: i18n.t('Dashboard') },
    [DATA_SET_REPORT_VIEW]: {
        value: DATA_SET_REPORT_VIEW,
        label: i18n.t('Data set report'),
    },
}

export default EVENT_TYPES
