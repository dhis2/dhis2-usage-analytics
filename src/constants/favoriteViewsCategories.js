import i18n from '@dhis2/d2-i18n'

export const SUM_ALL = 'SUM_ALL'
export const SUM_TOTAL = 'SUM_TOTAL'
export const AVEGAGE_ALL = 'AVEGAGE_ALL'
export const AVERAGE_TOTAL = 'AVERAGE_TOTAL'

const FAVORITE_VIEWS_CATEGORIES = [
    { id: SUM_ALL, displayName: i18n.t('All favorite views') },
    { id: SUM_TOTAL, displayName: i18n.t('Total views') },
    {
        id: AVEGAGE_ALL,
        displayName: i18n.t('All favorite average views per user'),
    },
    { id: AVERAGE_TOTAL, displayName: i18n.t('Total views per user') },
]
export default FAVORITE_VIEWS_CATEGORIES
