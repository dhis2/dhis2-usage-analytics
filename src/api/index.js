import { TOP_FAVORITES } from '../constants/categories'
import { getJSON } from './get'

const api = (() => {
    let baseUrl = null

    const setBaseUrl = url => {
        baseUrl = url
    }

    const getFavorites = ({ eventType, pageSize, sortOrder }) => {
        const queryParams = `eventType=${eventType}&pageSize=${pageSize}&sortOrder=${sortOrder}&_=${Date.now()}`
        return getJSON(baseUrl, `dataStatistics/favorites?${queryParams}`)
    }

    const getDataStatistics = ({ startDate, endDate, interval }) => {
        const queryParams = `startDate=${startDate}&endDate=${endDate}&interval=${interval}&_=${Date.now()}`
        return getJSON(baseUrl, `dataStatistics?${queryParams}`)
    }

    const getUsageData = filter => {
        return filter.category === TOP_FAVORITES
            ? getFavorites({
                  eventType: filter.eventType,
                  pageSize: filter.pageSize,
                  sortOrder: filter.sortOrder,
              })
            : getDataStatistics({
                  startDate: filter.startDate,
                  endDate: filter.endDate,
                  interval: filter.interval,
              })
    }

    return { setBaseUrl, getFavorites, getDataStatistics, getUsageData }
})()

export default api
