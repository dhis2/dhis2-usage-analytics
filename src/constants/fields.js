import * as CATS from './categories'
import * as FAV_VIEW_CATS from './favoriteViewsCategories'
import * as PROPS from './properties'

const FAVORITE_VIEWS_TOTAL_FIELDS = [
    PROPS.date,
    PROPS.mapViews,
    PROPS.chartViews,
    PROPS.pivotTableViews,
    PROPS.eventReportViews,
    PROPS.eventChartViews,
    PROPS.dashboardViews,
    PROPS.dataSetReportViews,
    PROPS.totalViews,
]

const FAVORITE_VIEWS_AVERAGE_FIELDS = [
    PROPS.date,
    PROPS.averageMapViews,
    PROPS.averageChartViews,
    PROPS.averagePivotTableViews,
    PROPS.averageEventReportViews,
    PROPS.averageEventChartViews,
    PROPS.averageDashboardViews,
    PROPS.averageViews,
]

export const FIELDS = {
    [CATS.FAVORITE_VIEWS]: {
        [FAV_VIEW_CATS.SUM_ALL]: FAVORITE_VIEWS_TOTAL_FIELDS,
        [FAV_VIEW_CATS.SUM_TOTAL]: FAVORITE_VIEWS_TOTAL_FIELDS,
        [FAV_VIEW_CATS.AVERAGE_ALL]: FAVORITE_VIEWS_AVERAGE_FIELDS,
        [FAV_VIEW_CATS.AVERAGE_TOTAL]: FAVORITE_VIEWS_AVERAGE_FIELDS,
    },
    [CATS.FAVORITES_SAVED]: [
        PROPS.date,
        PROPS.savedMaps,
        PROPS.savedCharts,
        PROPS.savedPivotTables,
        PROPS.savedEventReports,
        PROPS.savedEventCharts,
        PROPS.savedDashboards,
        PROPS.savedIndicator,
    ],
    [CATS.USERS]: [PROPS.date, PROPS.activeUsers, PROPS.users],
    [CATS.TOP_FAVORITES]: [
        PROPS.position,
        PROPS.name,
        PROPS.views,
        PROPS.id,
        PROPS.created,
    ],
    [CATS.DATA_VALUES]: [PROPS.date, PROPS.savedDataValues],
}
export default FIELDS
