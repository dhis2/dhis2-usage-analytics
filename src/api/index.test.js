/* eslint-disable import/namespace,no-import-assign */
import * as get from './get'
import api from './index'
import { TOP_FAVORITES } from '../constants/categories'

beforeAll(() => {
    get.getJSON = jest.fn()
    Date.now = jest.fn(() => 'timestamp')
    api.setBaseUrl('baseUrl')
})

afterAll(() => {
    jest.resetAllMocks()
})

describe('getFavorites', () => {
    it('calls getJSON with the correct URL and queryString', () => {
        const params = {
            eventType: 'eventType',
            pageSize: 'pageSize',
            sortOrder: 'sortOrder',
        }
        get.getJSON.mockImplementationOnce((_baseUrl, url) =>
            Promise.resolve(url)
        )

        return expect(api.getFavorites(params)).resolves.toMatchSnapshot()
    })
})

describe('getDataStatistics', () => {
    it('calls getJSON with the correct URL and queryString', () => {
        const params = {
            startDate: 'startDate',
            endDate: 'endDate',
            interval: 'interval',
        }
        get.getJSON.mockImplementationOnce((_baseUrl, url) =>
            Promise.resolve(url)
        )

        return expect(api.getDataStatistics(params)).resolves.toMatchSnapshot()
    })
})

describe('getUsageData', () => {
    it('calls getJSON with the correct URL and queryString for TOP_FAVORITES category', () => {
        const params = {
            category: TOP_FAVORITES,
            eventType: 'eventType',
            pageSize: 'pageSize',
            sortOrder: 'sortOrder',
        }
        get.getJSON.mockImplementationOnce((_baseUrl, url) =>
            Promise.resolve(url)
        )

        return expect(api.getUsageData(params)).resolves.toMatchSnapshot()
    })

    it('calls getJSON with the correct URL and queryString in other cases', () => {
        const params = {
            startDate: 'startDate',
            endDate: 'endDate',
            interval: 'interval',
        }
        get.getJSON.mockImplementationOnce((_baseUrl, url) =>
            Promise.resolve(url)
        )

        return expect(api.getUsageData(params)).resolves.toMatchSnapshot()
    })
})
