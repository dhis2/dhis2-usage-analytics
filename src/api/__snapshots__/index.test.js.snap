// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDataStatistics calls getJSON with the correct URL and queryString 1`] = `"dataStatistics?startDate=startDate&endDate=endDate&interval=interval&_=timestamp"`;

exports[`getFavorites calls getJSON with the correct URL and queryString 1`] = `"dataStatistics/favorites?eventType=eventType&pageSize=pageSize&sortOrder=sortOrder&_=timestamp"`;

exports[`getUsageData calls getJSON with the correct URL and queryString for TOP_FAVORITES category 1`] = `"dataStatistics/favorites?eventType=eventType&pageSize=pageSize&sortOrder=sortOrder&_=timestamp"`;

exports[`getUsageData calls getJSON with the correct URL and queryString in other cases 1`] = `"dataStatistics?startDate=startDate&endDate=endDate&interval=interval&_=timestamp"`;

exports[`getUserLocale calls getJSON with the correct URL and queryString 1`] = `"userSettings"`;
